pyro/contrib/conjugate/infer.py

```diff
@@ -0,0 +1,167 @@
+from functools import reduce
+from operator import mul
+
+import pyro.distributions as dist
+from pyro.distributions.util import sum_leftmost
+from pyro.poutine.messenger import Messenger
+from pyro.poutine.replay_messenger import ReplayMessenger
+
+
+def _make_cls(base, static_attrs, instance_attrs, parent_linkage=None):
+    r"""
+    Dynamically create classes named `_ + base.__name__`, which extend the
+    base class with other optional instance and class attributes, and have
+    a custom `.expand` method to propagate these attributes on expanded
+    instances.
+
+    :param cls base: Base class.
+    :param dict static_attrs: static attributes to add to class.
+    :param dict instance_attrs: instance attributes for initialization.
+    :param str parent_linkage: attribute in the parent class that holds
+        a reference to the distribution class.
+    :return cls: dynamically generated class.
+    """
+    def _expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(cls, _instance)
+        for attr in instance_attrs:
+            setattr(new, attr, getattr(self, attr))
+        if parent_linkage:
+            setattr(new.parent, parent_linkage, new)
+        return base.expand(self, batch_shape, _instance=new)
+
+    name = "_" + base.__name__
+    cls = type(name, (base,), instance_attrs)
+    for k, v in static_attrs.items():
+        setattr(cls, k, v)
+    cls.expand = _expand
+    return cls
+
+
+def _latent(base, parent):
+    return _make_cls(base, {"collapsible": True}, {"site_name": None, "parent": parent}, "_latent")
+
+
+def _conditional(base, parent):
+    return _make_cls(base, {"marginalize_latent": True}, {"parent": parent}, "_conditional")
+
+
+def _compound(base, parent):
+    return _make_cls(base, {}, {"parent": parent})
+
+
+class BetaBinomialPair(object):
+    def __init__(self):
+        self._latent = None
+        self._conditional = None
+
+    def latent(self, *args, **kwargs):
+        self._latent = _latent(dist.Beta, parent=self)(*args, **kwargs)
+        return self._latent
+
+    def conditional(self, *args, **kwargs):
+        self._conditional = _conditional(dist.Binomial, parent=self)(*args, **kwargs)
+        return self._conditional
+
+    def posterior(self, obs):
+        concentration1 = self._latent.concentration1
+        concentration0 = self._latent.concentration0
+        total_count = self._conditional.total_count
+        reduce_dims = len(obs.size()) - len(concentration1.size())
+        num_obs = reduce(mul, obs.size()[:reduce_dims], 1)
+        total_count = total_count[tuple(slice(1) if i < reduce_dims else slice(None)
+                                  for i in range(total_count.dim()))].reshape(concentration0.shape)
```

Comment:
could you add a comment in the code about the purpose of this computation?

Reply:
This is basically trying to unexpand `total_count` to have the same shape as `concentration0`. I have added a comment and also made it so that it will throw an error if this unexpand fails. I have also changed the test to have non-uniform `total_count`.